### PR TITLE
Backend request tracing

### PR DESCRIPTION
**Pull request by elliotcm**
Use perl to set GOVUK-Request-Id if not set.

Tested as "installs nginx with perl and the config doesn't make nginx angry" on backend, frontend, api and licensify load balancer VMs.

### CODE REVIEW

**Comment by rjw1** [comment]
yay! :dancers: :dancers: :party: :eggplant:

**Comment by boffbowsh** [comment]
giphy

**Comment by boffbowsh** [comment]
If we're going to add intelligence to nginx to provide a request id, maybe it should be a uuid rather than a pseudo-random string of identifiable request data. http://ryandlane.com/blog/2014/12/11/using-lua-in-nginx-for-unique-request-ids-and-millisecond-times-in-logs/

**Comment by elliotcm** [comment]
@boffbowsh I wanted to avoid going down the path of adding perl or lua modules at this point, neither languages have UUIDs in their stdlib.

**Comment by boffbowsh** [comment]
Fair enough 👍

**Comment by elliotcm** [comment]
@rjw1 / @boffbowsh Are you OK with me merging this?

File: modules/loadbalancer/templates/nginx_balance.conf.erb [hunk 1]
```diff
@@ -18,10 +18,25 @@ server {
...
+    my $r = shift;
+    my $current_header = $r->header_in("GOVUK-Request-Id");
+
+    if (defined $current_header && $current_header ne "") {
+      return $current_header;
+    } else {
+      my $pid = $r->variable("pid");
+      my $msec = $r->variable("msec");
+      my $remote_addr = $r->variable("remote_addr");
+      my $request_length = $r-variable("request_length");
```
Comment (steventux) [inline modules/loadbalancer/templates/nginx_balance.conf.erb]: shouldn't this be:
`my $request_length = $r->variable("request_length");`
assuming it's de-referencing `request_length` from `$r` and not some other tricksy Perl thing?

Comment (elliotcm) [inline modules/loadbalancer/templates/nginx_balance.conf.erb]: You are right, will fix

**Comment by elliotcm** [comment]
@steventux updated

**Comment by jabley** [comment]
And so we go full circle :sob: https://github.com/gds-attic/pp-puppet/pull/586

**Pull request merged**